src/Saritasa.Tools.Messages/Common/Expressions/Compilation/CompiledExpressionProvider.cs

```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Saritasa.Tools.Messages.Common.Expressions.Compilation
+{
+    public class CompiledExpressionProvider : ICompiledExpressionProvider
+    {
+        private ConcurrentDictionary<MethodInfo, dynamic> cache = new ConcurrentDictionary<MethodInfo, dynamic>();
+
+        public IReadOnlyDictionary<MethodInfo, dynamic> KeyValues => cache;
+
+        public Func<TInput, TResult> GetOrAdd<TInput, TResult>(MethodInfo methodInfo, Func<dynamic> factory)
```

Comment:
@krasninja is there need for Action delegate with parameters ? 
And I'll add more generic parameters for cache.